### DISCUSSION TITLE
Multi-tile entities and disabling this crate rendering certain tiles on a per-entity basis

### DISCUSSION THREAD

**Discussion by Veritius** [StarArawn/bevy_ecs_tilemap#304] [Q&A] [answered]
I'm wondering about three main things right now.
1. Is there support or planned support for multi-tile entities, aka single entities that take up multiple spaces on the tile?
2. Is there a way to stop this crate rendering specific entities? Quickly skimming the documentation on docs.rs makes me think this is done by not having `bevy_ecs_tilemap::tiles::TileTexture` on the entity.

**Comment by bzm3r** [comment, marked as answer]
1. Yes, there is support for this. A `TilemapBundle` has two components: `TilemapGridSize`, and `TilemapTileSize`. `TilemapGridSize` tells you the spacing between grid points, while `TilemapTileSize` specifies the size of the tiles to be placed at each grid point. If you look at `examples/animation.rs`, you will see how the [flower tiles are bigger](https://github.com/StarArawn/bevy_ecs_tilemap/blob/eabd66eac4c5204ec54646141ffb26935cace88d/examples/animation.rs#L30) (twice the size) of the [background grass tiles](https://github.com/StarArawn/bevy_ecs_tilemap/blob/eabd66eac4c5204ec54646141ffb26935cace88d/examples/animation.rs#L23), but the flowers share the same grid size as the background, so that we can place them relative to the background tiles.

Note however that every Tilemap can only contain tiles of the same size. So like the example `animation.rs`, you will want to have one Tilemap "layer" for each tile size: [flowers](https://github.com/StarArawn/bevy_ecs_tilemap/blob/eabd66eac4c5204ec54646141ffb26935cace88d/examples/animation.rs#L68) and [background](https://github.com/StarArawn/bevy_ecs_tilemap/blob/eabd66eac4c5204ec54646141ffb26935cace88d/examples/animation.rs#L34).

2. You can toggle whether a tile is rendered or not using visibility. See the example [`examples/visibility.rs`](https://github.com/StarArawn/bevy_ecs_tilemap/blob/eabd66eac4c5204ec54646141ffb26935cace88d/examples/visibility.rs#L71)

**Reply by Veritius** [reply]
>Note however that every Tilemap can only contain tiles of the same size. 

Unfortunately, I was more looking for differently sized entities being able to be placed on the same grid, much like Godot. I'm sure I can work something out, however.

**Reply by StarArawn** [reply]
> >Note however that every Tilemap can only contain tiles of the same size. 
> 
> Unfortunately, I was more looking for differently sized entities being able to be placed on the same grid, much like Godot. I'm sure I can work something out, however.

You might be better off just using bevy sprites if you have different sizes.

**Reply by Veritius** [reply]
Sprites unfortunately would not be good enough. I plan to use a grid based construction system, where smaller structures like floors are a single tile, but larger machines like generators are several tiles. I wanted to avoid any chance of overlapping or other such issues, and multiple different layers for very different machines is not an acceptable solution for my purposes.

**Reply by bzm3r** [reply]
@Veritius Well, the reason why we do not allow tiles of different sizes within the same tilemap is in order to improve rendering performance. A tilemap can effectively be thought of as something that is "batched" in a particular way for the GPU. 

It should not be confused for a logical separation between entities (although often times, it ends up working out that way too). It is a rendering separation. Therefore, do not use `TilemapTileSize` as a way to check collision between objects. Instead, you should have a `MachineSize` type which gets converted into a `TilemapTileSize` based on the size of the machine. 

In other words, your logical separation between entities will probably have to be maintained using standard bevy techniques: i.e. using marker components and so on. For my use cases, I tend to implement a trait which allows me to create the right `TileBundle` for a given entity given the components on it (I do this during entity creation time).

**Reply by Veritius** [reply]
> @Veritius Well, the reason why we do not allow tiles of different sizes within the same tilemap is in order to improve rendering performance. A tilemap can effectively be thought of as something that is "batched" in a particular way for the GPU.
> 
> It should not be confused for a logical separation between entities (although often times, it ends up working out that way too). It is a rendering separation. Therefore, do not use `TilemapTileSize` as a way to check collision between objects. Instead, you should have a `MachineSize` type which gets converted into a `TilemapTileSize` based on the size of the machine.
> 
> In other words, your logical separation between entities will probably have to be maintained using standard bevy techniques: i.e. using marker components and so on. For my use cases, I tend to implement a trait which allows me to create the right `TileBundle` for a given entity given the components on it (I do this during entity creation time).

Thank you, that's very interesting. I'll have a more in depth look later on today.
